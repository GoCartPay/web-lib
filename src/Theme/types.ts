/** @file Types used in theming */

/** A single unit in the design token, ie: the value of `color.primary.main` */
export interface TokenUnit<Value = string> {
  description?: string // eg: "Reflects the text.primary variable from the theme object",
  type: 'color' | 'custom-fontStyle' | string
  value: Value // string | Record<string, string | number> // eg: "#000000de" or {"fontSize": 12},
  extensions: {
    'org.lukasoppermann.figmaDesignTokens': {
      styleId: string // eg: "S:3954b6510df8a66fae1298393ec2109cd4f52516,",
      exportKey: 'color' | 'font' | string
    }
  }
}

type ColorUnit = TokenUnit<string>

type TypographyUnit = TokenUnit<{
  fontSize: number // eg: 96,
  textDecoration: string // eg: 'none'
  fontFamily: string // eg: 'TWK Lausanne'
  fontWeight: number // eg: 400
  fontStyle: string // eg: 'normal'
  fontStretch: string // eg: 'normal'
  letterSpacing: number // eg: -2
  lineHeight: number // eg: 108
  paragraphIndent: number // eg: 0
  paragraphSpacing: number // eg: 0
  textCase: string // eg: 'none'
}>

export type ElevationUnitValue = {
  shadowType: string // eg: "dropShadow",
  radius: number // eg: 64,
  color: string // eg: "#10182824",
  offsetX: number // eg: 0,
  offsetY: number // eg: 32,
  spread: number // eg: -12
}

/** An elevation unit with one light source */
export type SingleElevationUnit = TokenUnit<ElevationUnitValue>

/** An elevation unit with two light sources */
export type DoubleElevationUnit = {
  '0': { type: string; value: ElevationUnitValue }
  '1': { type: string; value: ElevationUnitValue }
  description: TokenUnit['description']
  extensions: TokenUnit['extensions']
}

export type ElevationUnit = SingleElevationUnit | DoubleElevationUnit

/**
 * Our base-level design token. Note that the actual design token
 * has far more data; this is the only data we use
 **/
export interface DesignToken {
  color: {
    text: {
      primary: ColorUnit
      secondary: ColorUnit
      disabled: ColorUnit
    }
    primary: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast_text: ColorUnit
    }
    secondary: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast: ColorUnit
    }
    error: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast: ColorUnit
    }
    warning: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast: ColorUnit
    }
    info: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast: ColorUnit
    }
    success: {
      main: ColorUnit
      dark: ColorUnit
      light: ColorUnit
      contrast: ColorUnit
    }
  }
  font: {
    typography: {
      title: TypographyUnit
      h1: TypographyUnit
      h2: TypographyUnit
      h3: TypographyUnit
      h4: TypographyUnit
      h5: TypographyUnit
      h6: TypographyUnit
      'body 1': TypographyUnit
      'body 2': TypographyUnit
      'subtitle 1': TypographyUnit
      'subtitle 2': TypographyUnit
      overline: TypographyUnit
      caption: TypographyUnit
      legal: TypographyUnit
      legal_url: TypographyUnit
    }
  }
  effect: {
    elevation: {
      1: ElevationUnit
      2: ElevationUnit
      3: ElevationUnit
      4: ElevationUnit
      5: ElevationUnit
      6: ElevationUnit
      7: ElevationUnit
    }
  }
}
