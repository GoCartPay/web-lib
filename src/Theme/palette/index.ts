/**
 * @file Process the design token's color settings into something ready to
 * use with MUI's `createTheme`
 **/

import design from '../design-tokens.tokens.json'
import { DesignToken, TokenUnit } from '../types'
import { Palette } from '@mui/material'

// Type imported JSON
const token: DesignToken = design

/** Pull value out of token object */
const mapColor = (color: { [token: string]: TokenUnit }) =>
  Object.entries(color).reduce(
    (acc, cur) => ({ ...acc, [cur[0]]: cur[1].value }),
    {}
  )

/** Palette, ready to be added to theme or used on it's own */
export const palette: Partial<Palette> = Object.keys(token.color).reduce(
  (acc: Partial<Palette>, cur: keyof DesignToken['color']) => ({
    ...acc,
    [cur]: mapColor(token.color[cur])
  }),
  {}
)
