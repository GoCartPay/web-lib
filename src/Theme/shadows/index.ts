/**
 * @file Process the design token's `effect.elevation` settings into
 * something ready to use with MUI's `createTheme`
 **/

import design from '../design-tokens.tokens.json'
import { createTheme } from '@mui/material/styles'
import type { Shadows } from '@mui/material/styles/shadows'
import type {
  DesignToken,
  ElevationUnitValue,
  SingleElevationUnit,
  DoubleElevationUnit
} from '../types'

// Type imported JSON
const token: DesignToken = design

const defaultTheme = createTheme({})

const customElevationLevels = ['1', '2', '3', '4', '5', '6', '7']

/** Given an object of CSS rules, convert into box-shadow syntax */
const convertValueToStyle = (v: ElevationUnitValue) =>
  /* offset-x | offset-y | blur-radius | spread-radius | color */
  `${v.offsetX}px ${v.offsetY}px ${v.radius}px ${v.spread}px ${v.color}`

/** Box shadows, ready to be added to theme or used on it's own */
const customShadows = Object.keys(token.effect.elevation)
  .sort()
  .reduce(
    (acc, cur: '1' | '2' | '3' | '4' | '5' | '6' | '7') => {
      if (!customElevationLevels.includes(cur)) {
        return acc
      }

      if (
        Object.prototype.hasOwnProperty.call(
          token.effect.elevation[cur],
          'value'
        )
      ) {
        const valueObj = token.effect.elevation[cur] as SingleElevationUnit
        return [...acc, convertValueToStyle(valueObj.value)]
      } else {
        const valueObj = token.effect.elevation[cur] as DoubleElevationUnit
        return [
          ...acc,
          `${convertValueToStyle(valueObj['0'].value)}, ${convertValueToStyle(
            valueObj['1'].value
          )}`
        ]
      }
    },
    ['none']
  )

export const shadows = customShadows.concat(
  [...defaultTheme.shadows].slice(customShadows.length)
) as Shadows
