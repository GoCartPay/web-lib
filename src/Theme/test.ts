import theme from './index'

// Note that if the design token changes a used value, the snapshot will need to be updated
it('Produces correct output', () => {
  expect(theme).toMatchSnapshot()
})
