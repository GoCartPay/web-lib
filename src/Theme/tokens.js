const tokens = {
    shape: {
        borderRadius: '20px'
    }
}

export { tokens }
