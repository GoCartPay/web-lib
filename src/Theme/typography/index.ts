/**
 * @file Process the design token's `font.typography` settings into
 * something ready to use with MUI's `createTheme`
 **/

import design from '../design-tokens.tokens.json'
import type { DesignToken } from '../types'

// Allow custom variants in TS
declare module '@mui/material/styles' {
  interface TypographyVariants {
    legal: React.CSSProperties
    legal_url: React.CSSProperties
  }

  // Allow configuration using `createTheme`
  interface TypographyVariantsOptions {
    legal?: React.CSSProperties
    legal_url?: React.CSSProperties
  }
}

// Update the Typography's variant prop options
declare module '@mui/material/Typography' {
  interface TypographyPropsVariantOverrides {
    legal: true
    legal_url: true
  }
}

// Type imported JSON
const token: DesignToken = design

/** Typography, ready to be added to theme or used on it's own */
export const typography = Object.keys(token.font.typography).reduce(
  (acc, cur: keyof DesignToken['font']['typography']) => ({
    ...acc,
    [cur.replace(/ /g, '')]: token.font.typography[cur].value
  }),
  { fontFamily: 'TWK Lausanne' }
)
