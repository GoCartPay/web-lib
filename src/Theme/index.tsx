/** @file Main theme utility */
import React from 'react'
import {
  createTheme,
  ThemeProvider as MuiThemeProvider
} from '@mui/material/styles'
import { palette } from './palette'
import { shadows } from './shadows'
import { typography } from './typography'
import TWKLausanne from './typography/fonts/TWK-Lausanne/TWKLausanne-400.ttf'
import CssBaseline from '@mui/material/CssBaseline'

/** Our partial theme */
const mappedTheme = {
  palette,
  shadows,
  typography
}

/** Our full theme */
export const theme = createTheme({
  ...mappedTheme,
  components: {
    MuiCssBaseline: {
      styleOverrides: `
      @font-face {
        font-family: 'TWK Lausanne';
        font-style: normal;
        font-display: swap;
        font-weight: 400;
        src: url(${TWKLausanne}) format('truetype');
        unicodeRange: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF;
      }
    `
    }
  }
})

console.log({ theme })

/** Theme provider, pre-filled with Spruce theme */
export const ThemeProvider = (props?: Record<string, unknown>) => {
  const { children, ...otherProps } = props
  return (
    <MuiThemeProvider theme={theme} {...otherProps}>
      <CssBaseline />
      {children}
    </MuiThemeProvider>
  )
}

export default theme
