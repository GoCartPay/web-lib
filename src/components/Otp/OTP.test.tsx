import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import OTP from './';
import { ThemeProvider } from '../../Theme';
const mockOnChange = jest.mock;

describe('OTP renders and works correctly', () => {
  it('renders with expected content and does not crash', () => {
    const { container } = render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
        />
      </ThemeProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('will render an amount of input fields equal to the number defined in codeLength', () => {
    render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
          codeLength={4}
        />
      </ThemeProvider>
    );
    const inputLength = screen.getAllByRole('textbox').length;
    expect(inputLength).toEqual(4);
  });

  it('when otp is rendered, active styling should be applied and first input should have focus', () => {
    const { container } = render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
        />
      </ThemeProvider>
    );

    expect(container.firstChild).toHaveStyle('border: solid 1px #117B74CC');

    const firstOtpInput = screen.getByRole('textbox', { name: /otp-field-0/});
    
    expect(firstOtpInput).not.toHaveAttribute('placeholder');
    expect(firstOtpInput).toHaveFocus();
  });

  it('will conditionally render error styling when there is an error', () => {
    const { container } = render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
          hasError
        />
      </ThemeProvider>
    );
    expect(container.firstChild).toHaveStyle('border: solid 1px #DF2113')
  });

  it('will conditionally render isComplete styling when OTP has completed', () => {
    const { container } = render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
          isComplete
        />
      </ThemeProvider>
    );
    expect(container.firstChild).toHaveStyle('border: 1px solid rgb(42, 208, 98)');
  });

  it('will conditionally render error styling when OTP has an error', () => {
    const { container } = render(
      <ThemeProvider>
        <OTP
          onChange={mockOnChange}
          value={''}
          hasError
        />
      </ThemeProvider>
    );
    expect(container.firstChild).toHaveStyle('border: solid 1px #DF2113');
  });

  it('when a user types into OTP field, the focus field should be updated and the following input whould become active', () => {
    render(
      <ThemeProvider>
        <OTP
          onChange={(value) => console.log(value)}
          value={''}
          hasError
          codeLength={3}
        />
      </ThemeProvider>
    );
    
    const firstOtpInput = screen.getByRole('textbox', { name: /otp-field-0/});
    const secondOtpInput = screen.getByRole('textbox', { name: /otp-field-1/});
    const thirdOtpInput = screen.getByRole('textbox', { name: /otp-field-2/});

    userEvent.type(firstOtpInput, '1');
    expect(firstOtpInput).not.toHaveFocus();
    expect(secondOtpInput).not.toHaveAttribute('placeholder');
    expect(secondOtpInput).toHaveFocus();

    userEvent.type(secondOtpInput, '1');
    expect(firstOtpInput).not.toHaveFocus();
    expect(secondOtpInput).not.toHaveFocus();
    expect(thirdOtpInput).toHaveFocus();
    expect(thirdOtpInput).not.toHaveAttribute('placeholder');
  });

  it('keyboard events fire correctly (arrow right/left, backspace)', () => {
    render(
      <ThemeProvider>
        <OTP
          onChange={(value) => console.log(value)}
          value={''}
          hasError
          codeLength={3}
        />
      </ThemeProvider>
    );
    
    const firstOtpInput = screen.getByRole('textbox', { name: /otp-field-0/});
    const secondOtpInput = screen.getByRole('textbox', { name: /otp-field-1/});
    const thirdOtpInput = screen.getByRole('textbox', { name: /otp-field-2/});

    userEvent.type(firstOtpInput, '1');
    userEvent.type(secondOtpInput, '1');

    userEvent.keyboard('{arrowleft}');
    expect(secondOtpInput).toHaveFocus();

    userEvent.keyboard('{arrowright}');
    expect(thirdOtpInput).toHaveFocus();
    
    userEvent.keyboard('{backspace}');
    expect(secondOtpInput).toHaveFocus();

    userEvent.keyboard('{backspace}');
    expect(firstOtpInput).toHaveFocus();
    
  });
 
});