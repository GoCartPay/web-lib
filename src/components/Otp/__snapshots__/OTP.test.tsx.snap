// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTP renders and works correctly renders with expected content and does not crash 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1rik2fx-MuiPaper-root-Otp"
>
  <div
    class="MuiBox-root css-afxx7u"
  >
    <input
      aria-label="otp-field-0"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-0"
      pattern="[0-9]*"
      value=""
    />
    <input
      aria-label="otp-field-1"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-1"
      pattern="[0-9]*"
      placeholder="―"
      value=""
    />
    <input
      aria-label="otp-field-2"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-2"
      pattern="[0-9]*"
      placeholder="―"
      value=""
    />
    <input
      aria-label="otp-field-3"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-3"
      pattern="[0-9]*"
      placeholder="―"
      value=""
    />
    <input
      aria-label="otp-field-4"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-4"
      pattern="[0-9]*"
      placeholder="―"
      value=""
    />
    <input
      aria-label="otp-field-5"
      inputmode="numeric"
      maxlength="1"
      name="otp-field-5"
      pattern="[0-9]*"
      placeholder="―"
      value=""
    />
  </div>
</div>
`;
